Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -944,6 +944,387 @@ PhysicalParticleContainer::AddPlasma (int lev, RealBox part_realbox)
     // The function that calls this is responsible for redistributing particles.
 }
 
+void
+PhysicalParticleContainer::AddPlasmaFlux (int lev, amrex::Real dt)
+{
+    WARPX_PROFILE("PhysicalParticleContainer::AddPlasmaFlux()");
+
+    const Geometry& geom = Geom(lev);
+    const amrex::RealBox& part_realbox = geom.ProbDomain();
+
+    amrex::Real num_ppc_real = plasma_injector->num_particles_per_cell_real;
+#ifdef WARPX_DIM_RZ
+    Real rmax = std::min(plasma_injector->xmax, geom.ProbDomain().hi(0));
+#endif
+
+    const auto dx = geom.CellSizeArray();
+    const auto problo = geom.ProbLoArray();
+
+    Real scale_fac = 0._rt;
+    if (dx[plasma_injector->flux_normal_axis]*num_ppc_real != 0._rt) {
+        scale_fac = AMREX_D_TERM(dx[0], *dx[1], *dx[2])/dx[plasma_injector->flux_normal_axis]/num_ppc_real;
+    }
+
+    defineAllParticleTiles();
+
+    amrex::LayoutData<amrex::Real>* cost = WarpX::getCosts(lev);
+
+    const int nlevs = numLevels();
+    static bool refine_injection = false;
+    static Box fine_injection_box;
+    static int rrfac = 1;
+    // This does not work if the mesh is dynamic.  But in that case, we should
+    // not use refined injected either.  We also assume there is only one fine level.
```

Comment:
Do you like to add an `amrex::Abort()` here to guard that case?

Reply:
This bit of code was mostly copied from AddPlasma. Here "dynamic" presumably means dynamic mesh refinement which is not supported by WarpX anyway.

Reply:
Oh, how much do we copy? Is useful to put this in a function?

Reply:
When writing this new routine, I copied AddPlasma to start with and went through and made all of the needed changes. There is a fair amount of copied code left, but it's spread out in bits and pieces. There are a number of places that are grabbing local references to needed data, for example `p_ion_level`, that couldn't easily be moved into a separate subroutine. I agree that avoiding code duplication is a good thing, but I think it would be a force fit here and not necessarily help.